### PR TITLE
Eliminate the `Worker` State Reconciler

### LINKED ISSUE #8488: [GEP-22] Eliminate the `Worker` State Reconciler
**How to categorize this issue?**

/area control-plane-migration
/kind enhancement

**What would you like to be added**:
On our landscapes we hit a case where gardenlet was no longer able to update the Worker resource because we hit the etcd's default request limit of 1.5MiB:
```
task "Configuring shoot worker pools" failed: retry failed with context deadline exceeded, last error: etcdserver: request is too large
```

https://github.com/gardener/gardener/pull/8482 was a band-aid fix to mitigate the issue in short-term.

For long-term, @rfranzke shared that there is already the following TODO https://github.com/gardener/gardener/blob/7fbf64f49acec19cbd8398c8121830705db5388e/extensions/pkg/controller/worker/state_reconciler.go#L41-L42 which depends on this TODO (https://github.com/gardener/gardener/blob/7fbf64f49acec19cbd8398c8121830705db5388e/pkg/gardenlet/controller/shoot/shoot/reconciler_migrate.go#L179-L180)

@rfranzke shared:
> The idea is to not update the `.status.state` in the Worker field at all - gardenlet just fetches the machine state during migration phase and populates it directly to the ShootState (without the detour of first populating it to the Worker).

It is debatable whether this will permanently fix the problem or move it from the Worker resource to the ShootState resource. Still, it will be a nice improvement as CPM is a rare operation and the Shoot reconciliation won't be bothered by similar issue.

See also https://github.com/gardener/gardener/blob/e0ca8e718b56697de2e873d4383796f0b8b0dc82/docs/proposals/22-improved-usage-of-shootstate-api.md#eliminating-the-worker-state-reconciler

**Why is this needed**:
See above.

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area scalability cost
/kind enhancement

**What this PR does / why we need it**:
This PR drops the [`Worker` state reconciler](https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/worker/state_reconciler.go) that currently is [auto-added](https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/worker/controller.go#L75) to all `Worker` controllers. Please see [GEP-22](https://github.com/gardener/gardener/blob/e0ca8e718b56697de2e873d4383796f0b8b0dc82/docs/proposals/22-improved-usage-of-shootstate-api.md#eliminating-the-worker-state-reconciler) for information about the motivation.

With this PR, the migration flow is adapted as follows:
- `gardenlet` collects machine state and persists it in the `ShootState` under `.spec.gardener[].type=machine-state`.
- `gardenlet` shallow-deletes the machine-related resources (this was previously done by the [generic `Worker` actuator](https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/worker/genericactuator/actuator_migrate.go#L72-L99)).
- For a few releases, `gardenlet` no longer persists the state stored in `.status.state` of `Worker` resources (to not duplicate the machine state in the `ShootState`).

The restoration flow is adapted as follows:
- The generic `Worker` actuator tries to read the machine state directly from the `ShootState` in the garden cluster (which is now possible since #8001). This alleviates the necessity to let `gardenlet` read it and populate it somehow to the destination seed for the provider extensions to pick up - they can do it directly.
- `gardenlet` fetches the machine state from the `ShootState`'s `.spec.gardener[]` field (previously, the machine state was fetched from `.spec.extensions[].kind=Worker`) and still populates it to the `Worker`'s `.status.state` field. This is just to preserve backwards-compatibility, because registered extensions might not yet have vendored the newest library, hence they might still rely on this field during restoration. This will be dropped in a future version.

The PR also addresses a TODO leftover task from https://github.com/gardener/gardener/pull/8082 about only persisting the `ShootState` after all extension resources have been migrated.

**Which issue(s) this PR fixes**:
Fixes #8488

**Special notes for your reviewer**:
- The version skew policy for extensions was clarified as part of this PR.
- `github.com/gardener/machine-controller-manager` was revendored to `0.50.0` according to #8539. In this version, the legacy provider-specific API types are cleaned up. The revendoring was needed to make the `machine.sapcloud.io/v1alpha1` CRDs available for the integration tests.
- Accordingly, the no longer needed `MachineClass`-related methods were dropped from the generic `Worker` actuator interface.

/cc @plkokanov 

**Release note**:

```noteworthy operator
The `Worker` state reconciler has been dropped, i.e., updated provider extensions will no longer populate the machine state to the `.status.state` field of `Worker` resources. For a few releases, `gardenlet` will no longer persist any still existing data in the `.status.state` field of `Worker` resources during a control plane migration of a `Shoot`, and it will set `.status.state` to `nil` after a successful reconciliation or restore operation.
```
```breaking dependency
Provider extensions must now pass the `cluster.Cluster` object for the garden cluster to the `genericactuator.NewActuator` function. See [this](https://github.com/gardener/gardener/blob/8d2f116aa606e5181cd430e5063dd798629bdc78/cmd/gardener-extension-provider-local/app/app.go#L228-L246) for an example how to create such a `cluster.Cluster` object.
```
```noteworthy dependency
The `MachineClassKind()`, `MachineClass()`, and `MachineClassList()` methods have been dropped from the generic `Worker` actuator's interface and do not need to be implemented anymore.
```
```noteworthy developer
During the `Migrate` phase of a control plane migration of a `Shoot`, the state is now only persisted after all extension resources have been migrated. Consequently, make sure that you have added all state to the `.status.state` field of the respective extension object when running `Migrate()`.
```

### CODE REVIEW

File: pkg/utils/gardener/machines.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+const (
+	nameLabel = "name"
+	// MachineSetKind is the kind of the owner reference of a machine set
+	MachineSetKind = "MachineSet"
+	// MachineDeploymentKind is the kind of the owner reference of a machine deployment
+	MachineDeploymentKind = "MachineDeployment"
+)
+
+// BuildOwnerToMachinesMap builds a map from a slice of machinev1alpha1.Machine, that maps the owner reference
+// to a slice of machines with the same owner reference
```
Comment (timuthy) [inline pkg/utils/gardener/machines.go]: Nit
```suggestion
// BuildOwnerToMachinesMap returns a map that associates `MachineSet` names to the given `machines`.
```

File: pkg/utils/gardener/machines.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+			if machineDeploymentName, ok := machine.Labels[nameLabel]; ok {
+				ownerToMachines[machineDeploymentName] = append(ownerToMachines[machineDeploymentName], machines[index])
+			}
+		}
+	}
+	return ownerToMachines
+}
+
+// BuildOwnerToMachineSetsMap builds a map from a slice of machinev1alpha1.MachineSet, that maps the owner reference
+// to a slice of MachineSets with the same owner reference
```
Comment (timuthy) [inline pkg/utils/gardener/machines.go]: Nit
```suggestion
// BuildOwnerToMachineSetsMap returns a map that associates `MachineDeployment` names to the given `machineSets`.
```

File: pkg/utils/gardener/shootstate/machines.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+	machinev1alpha1 "github.com/gardener/machine-controller-manager/pkg/apis/machine/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gardenerutils "github.com/gardener/gardener/pkg/utils/gardener"
+)
+
+// MachineDeploymentState stores the last versions of the machine sets and machine to which the machine deployment
+// corresponds.
```
Comment (timuthy) [inline pkg/utils/gardener/shootstate/machines.go]:
```suggestion
// MachineDeploymentState stores the last versions of the machine sets and machines to which the machine deployment
// corresponds.
```

File: pkg/utils/gardener/shootstate/shootstate.go [hunk 4]
```diff
@@ -130,6 +130,36 @@ func computeGardenerData(
...
+
+	machineStateJSON, err := json.Marshal(machineState)
+	if err != nil {
+		return nil, fmt.Errorf("failed marshalling machine state to JSON: %w", err)
+	}
+
+	return append(secretsToPersist, gardencorev1beta1.GardenerResourceData{
+		Name: "machine-state",
+		Type: v1beta1constants.DataTypeMachineState,
+		Data: runtime.RawExtension{Raw: machineStateJSON},
```
Comment (timuthy) [inline pkg/utils/gardener/shootstate/shootstate.go]: We add secrets one-by-one to `gardener` and now all machine deployments under a single entry `machine-state`. Shouldn't we be consistent here?

File: pkg/component/extensions/worker/worker.go [hunk 5]
```diff
@@ -262,10 +262,28 @@ func (w *worker) deploy(ctx context.Context, operation string) (extensionsv1alph
 
 // Restore uses the seed client and the ShootState to create the Worker resources and restore their state.
 func (w *worker) Restore(ctx context.Context, shootState *gardencorev1beta1.ShootState) error {
+	// gardenlet persists the machine state in the ShootState's `.spec.gardener[]` list with `type=machine-state`.
+	// In order to use below function, we have to make the worker state appear under `.spec.extensions[]` list with
+	// the correct kind and name.
+	var (
+		shootStateCopy = shootState.DeepCopy()
```
Comment (timuthy) [inline pkg/component/extensions/worker/worker.go:273]: Can you explain why we need a copy here?

File: docs/extensions/migration.md [hunk 6]
```diff
@@ -73,18 +73,32 @@ In addition, extension controllers that use [referenced resources](referenced-re
...
 
-Most extension controller implementations follow a common pattern where a generic `Reconciler` implementation delegates to an `Actuator` interface that contains the methods `Reconcile` and `Delete`, provided by the extension. The two new methods `Migrate` and `Restore` have been added to all such `Actuator` interfaces, see [the infrastructure `Actuator` interface](https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/infrastructure/actuator.go) as an example. These methods are called by the generic reconcilers for the [migrate and restore operations](#migrate-and-restore-operations) respectively, and should be implemented by the extension according to the above guidelines.
+Most extension controller implementations follow a common pattern where a generic `Reconciler` implementation delegates to an `Actuator` interface that contains the methods `Reconcile` and `Delete`, provided by the extension.
+The two new methods `Migrate` and `Restore` have been added to all such `Actuator` interfaces, see [the infrastructure `Actuator` interface](https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/infrastructure/actuator.go) as an example.
```
Comment (timuthy) [inline docs/extensions/migration.md]:
```suggestion
Two methods `Migrate` and `Restore` are available in all such `Actuator` interfaces, see [the infrastructure `Actuator` interface](https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/infrastructure/actuator.go) as an example.
```

It's not so new anymore 😄

File: extensions/pkg/controller/worker/genericactuator/actuator_restore_test.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/gardener/shootstate"
+)
+
+var _ = Describe("ActuatorRestore", func() {
+	Describe("#addStateToMachineDeployment", func() {
+		var (
+			ctx              = context.TODO()
```
Comment (timuthy) [inline extensions/pkg/controller/worker/genericactuator/actuator_restore_test.go]:
```suggestion
		        ctx              = context.Background()
```

**Review by timuthy — COMMENTED** [review]
Nice PR and as usual very understandable 👏 😄 
Some minor points, rest looks good to me.

File: pkg/component/extensions/worker/worker.go (hunk 5 above)
Comment (rfranzke) [reply pkg/component/extensions/worker/worker.go:273]: A few lines below, I'm adding something to the `ShootState` which should only be visible in this function, i.e., I don't want to modify the passed object.

File: pkg/utils/gardener/shootstate/shootstate.go (hunk 4 above)
Comment (rfranzke) [reply pkg/utils/gardener/shootstate/shootstate.go]: We can think about refactoring the structure in the future. For now, the goal of this PR was just to move responsibilities from extensions to `gardenlet`. If I now start to also modify the structure, the PR will get even bigger, hence, I propose to look at this separately if really needed. The current structure (all in one object) has at least the benefit, that compressing the machine state should be easier (which is a task that we might need to follow up with, ref #8488).

**Review by timuthy — COMMENTED** [review]
/lgtm

**Review by timuthy — COMMENTED** [review]
/lgtm

**Review by timuthy — COMMENTED** [review]
/approve

**Pull request merged**